k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/SelectOrExamineResponse.java

```diff
@@ -1,52 +1,121 @@
+    private QresyncResponse qresyncResponse;
+    private Boolean readWriteMode;
+
+    private SelectOrExamineResponse() {
+    }
+
+    public static SelectOrExamineResponse parse(List<ImapResponse> imapResponses, ImapFolder folder) throws IOException,
+            MessagingException {
+        SelectOrExamineResponse selectOrExamineResponse = new SelectOrExamineResponse();
+        for (ImapResponse imapResponse : imapResponses) {
+            if (imapResponse.isTagged() || !equalsIgnoreCase(imapResponse.get(0), Responses.OK)
+                    || !imapResponse.isList(1)) {
+                continue;
+            }
+            handleUidValidity(imapResponse, selectOrExamineResponse);
+            handleHighestModSeq(imapResponse, selectOrExamineResponse);
+            handlePermanentFlags(imapResponse, selectOrExamineResponse, folder.store.getPermanentFlagsIndex());
+        }
+        selectOrExamineResponse.readWriteMode = isModeReadWrite(ImapUtility.getLastResponse(imapResponses));
+        if (folder.supportsQresync()) {
+            selectOrExamineResponse.qresyncResponse = QresyncResponse.parse(imapResponses, folder);
+        } else {
+            selectOrExamineResponse.qresyncResponse = null;
+        }
+        return selectOrExamineResponse;
+    }
+
+    private static void handleUidValidity(ImapResponse imapResponse, SelectOrExamineResponse selectOrExamineResponse) {
+        ImapList responseTextList = imapResponse.getList(1);
+        if (responseTextList.size() < 2 || !(equalsIgnoreCase(responseTextList.get(0), Responses.UIDVALIDITY))) {
+            return;
+        }
+        selectOrExamineResponse.uidValidity = Long.parseLong(responseTextList.getString(1));
+    }
 
-    private SelectOrExamineResponse(Boolean readWriteMode) {
-        this.readWriteMode = readWriteMode;
+    private static void handleHighestModSeq(ImapResponse imapResponse, SelectOrExamineResponse selectOrExamineResponse)
+            throws IOException, MessagingException {
+        Long highestModSeq = ImapUtility.extractHighestModSeq(imapResponse);
+        if (highestModSeq != null) {
+            selectOrExamineResponse.highestModSeq = highestModSeq;
+        }
     }
 
-    public static SelectOrExamineResponse parse(ImapResponse response) {
-        if (!response.isTagged() || !equalsIgnoreCase(response.get(0), Responses.OK)) {
+    private static void handlePermanentFlags(ImapResponse imapResponse, SelectOrExamineResponse selectOrExamineResponse,
+            Set<Flag> permanentFlags) {
+        PermanentFlagsResponse permanentFlagsResponse = PermanentFlagsResponse.parse(imapResponse);
+        if (permanentFlagsResponse == null) {
+            return;
+        }
+
+        permanentFlags.addAll(permanentFlagsResponse.getFlags());
+        selectOrExamineResponse.canCreateKeywords = permanentFlagsResponse.canCreateKeywords();
+    }
+
+    private static Boolean isModeReadWrite(ImapResponse imapResponse) {
```

Comment:
it's not obvious to me under what circumstances this might return null, as opposed to false. if it happens when the response doesn't contain a valid indicator, should that ever happen? if you think this is the right signature for this method, the name should reflect that failure case, e.g. by adding `IfAvailable`